memtier, controllers: split out dynamic page demotion into a controller.

klihub opened the pull request:
This patch series is the first step in generalizing and improving dynamic page demotion by
introducing a new `page-migration` controller and splitting out the core of the functionality
from the `memtier` policy to this new controller. With this patch series applied any policy
can activate page demotion for container by simply attaching demotion policy decisions to
the desired containers.

In particular, the patch series
- adds functions for getting/setting page-migration policy/preferences to cache.Container
- adds a new `page-migration` controller
- moves the core of the demotion functionality to the new controller
- only leaves demotion/migration policy decision making in `memtier`
- fixes `resource-manager` to execute controller post-stop hooks for deleted containers

The current page-migration policy decision is implicitly 'idle page migration', with the policy
only providing the set of source and target nodes. Detecting idle pages and migrating them
from sources to targets is entirely taken care of in the page-migration controller. This should
be fairly easy to generalize in the future.

marquiz commented on pkg/cri/resource-manager/cache/cache.go:
```diff
@@ -446,6 +455,26 @@ type Device struct {
+	TargetNodes system.IDSet // should be migrated to these NUMA nodes
+}
+
+func (pm *PageMigrate) Clone() *PageMigrate {
```
nit: add a comment/docstring to the exported methods

klihub replied:
Added docstring.

marquiz commented on pkg/cri/resource-manager/cache/container.go:
```diff
@@ -811,6 +811,15 @@ func (c *container) GetToptierLimit() int64 {
 	return c.ToptierLimit
 }
 
+func (c *container) SetPageMigration(p *PageMigrate) {
```
nit: comment/docstring

klihub replied:
It is present in the interface definition. Ditto for GetPageMigration().

marquiz replied:
Yes, of course, sorry about this

marquiz commented on pkg/cri/resource-manager/control/page-migrate/flags.go:
```diff
@@ -0,0 +1,42 @@
+	"github.com/intel/cri-resource-manager/pkg/config"
+)
+
+// Options captures our conigurable controller parameters.
```
s/Options/options/

klihub replied:
Updated. And fixed the typo, too... ;-)

marquiz commented on pkg/cri/resource-manager/control/page-migrate/page-migrate.go:
```diff
@@ -0,0 +1,99 @@
+}
+
+// migrationError creates a memory I/O-controller-specific formatted error message.
+func migrationError(format string, args ...interface{}) error {
```
memory I/O-controller??

klihub replied:
Fixed.

marquiz commented on pkg/cri/resource-manager/control/page-migrate/demoter.go:
```diff
@@ -217,7 +246,7 @@ func resetDirtyBit(pid string) error {
 
 // ResetDirtyBit unsets soft-dirty bits for all processes in a container.
-func (d *demoter) ResetDirtyBit(c *container) error {
+func (d *demoter) resetDirtyBit(c *container) error {
```
nit: Change docstring correspondingly

marquiz commented on pkg/cri/resource-manager/control/page-migrate/demoter.go:
```diff
@@ -97,32 +98,69 @@ func copyPagePool(p pagePool) pagePool {
+	}
+}
+
+func (d *demoter) Stop() {
```
nit: I'd add comment to the exported methods. Or, alternatively, if demoter implements an interface state that in the demoter definition

klihub replied:
Added docstring to exported methods Stop() and Reconfigure().

marquiz reviewed (commented):
Looks good, just a few very small nits.

One thing that puzzled me is why do we track containers locally in the controller? 🤔

klihub commented:
> Looks good, just a few very small nits.
> 
> One thing that puzzled me is why do we track containers locally in the controller?

To avoid having to serialize access to the cache against request and event processing,
IOW against the resource-manager and policies. Or to put it another way to avoid having
to perform the potentially lengthy process of page scanning with the cache locked (which
then blocks request processing). 

An alternative would have been to duplicate what we had in the policy:
  - introduce controller events (basically callbacks invoked by the resource-manager with the cache locked)
  - periodically inject a controller event when we want to scan pages
  - scan pages from the event callback, with resource-manager holding the lock for us

Instead we now track a bare minimal copy of cached containers, just enough for doing
page scanning and migration completely on our own.

marquiz commented:
> To avoid having to serialize access to the cache against request and event processing,
> IOW against the resource-manager and policies. Or to put it another way to avoid having
> to perform the potentially lengthy process of page scanning with the cache locked (which
> then blocks request processing).
> 
> An alternative would have been to duplicate what we had in the policy:
> 
> * introduce controller events (basically callbacks invoked by the resource-manager with the cache locked)
> * periodically inject a controller event when we want to scan pages
> * scan pages from the event callback, with resource-manager holding the lock for us
> 
> Instead we now track a bare minimal copy of cached containers, just enough for doing
> page scanning and migration completely on our own.

Thanks for the explanation, makes sense 👍 Do we have this documented somewhere? If not, I think it would be geed to add e.g. a short comment in the code explaining the reason we're doing this. Might make it easier to read/maintain code in the future. WDYT?

klihub commented:
> Thanks for the explanation, makes sense  Do we have this documented somewhere? If not, I think it would be geed to add e.g. a short comment in the code explaining the reason we're doing this. Might make it easier to read/maintain code in the future. WDYT?

Yes, that's very much true. I'll add a comment explaining this to the code.

klihub commented:
> Yes, that's very much true. I'll add a comment explaining this to the code.

Added a comment about why we copy a subset of the containers to the controller. While at it, also updated the code to use a set of functions mimicking the corresponding cache functions when dealing with locally cached containers.

marquiz reviewed (approved):
Wow, that comment is pretty exhaustive 😊 👍 

LGTM
